pkg/ebpftracer/c/tracee.bpf.c

```diff
@@ -6092,125 +5693,52 @@ statfunc u32 cgroup_skb_generic(struct __sk_buff *ctx, void *cgrpctxmap)
     if (!sk)
         return 1;
 
-    nethdrs hdrs = {0}, *nethdrs = &hdrs;
-
-    void *dest;
-
-    u32 size = 0;
-    u32 family = ctx->family;
-
-    switch (family) {
-        case PF_INET:
-            dest = &nethdrs->iphdrs.iphdr;
-            size = get_type_size(struct iphdr);
-            break;
-        case PF_INET6:
-            dest = &nethdrs->iphdrs.ipv6hdr;
-            size = get_type_size(struct ipv6hdr);
-            break;
-        default:
-            return 1; // verifier
-    }
-
-    // load layer 3 headers (for cgrpctxmap key/indexer)
-
-    if (bpf_skb_load_bytes_relative(ctx, 0, dest, size, 1))
-        return 1;
-
-    //
-    // IGNORE UNSUPPORTED PROTOCOLS, CREATE INDEXER TO OBTAIN EVENT
-    //
-
-    indexer_t indexer = {0};
-    indexer.ts = ctx->tstamp;
-
-    u32 ihl = 0;
-    switch (family) {
-        case PF_INET:
-            if (nethdrs->iphdrs.iphdr.version != 4) // IPv4
-                return 1;
-
-            ihl = nethdrs->iphdrs.iphdr.ihl;
-            if (ihl > 5) { // re-read IPv4 header if needed
-                size -= get_type_size(struct iphdr);
-                size += ihl * 4;
-                bpf_skb_load_bytes_relative(ctx, 0, dest, size, 1);
-            }
-
-            switch (nethdrs->iphdrs.iphdr.protocol) {
-                case IPPROTO_TCP:
-                case IPPROTO_UDP:
-                case IPPROTO_ICMP:
-                    break;
-                default:
-                    return 1; // unsupported proto
-            }
-
-            // add IPv4 header items to indexer
-            indexer.ip_csum = nethdrs->iphdrs.iphdr.check;
-            indexer.src.in6_u.u6_addr32[0] = nethdrs->iphdrs.iphdr.saddr;
-            indexer.dst.in6_u.u6_addr32[0] = nethdrs->iphdrs.iphdr.daddr;
-            break;
-
-        case PF_INET6:
-            // TODO: dual-stack IP implementation unsupported for now
-            // https://en.wikipedia.org/wiki/IPv6_transition_mechanism
-            if (nethdrs->iphdrs.ipv6hdr.version != 6) // IPv6
-                return 1;
-
-            switch (nethdrs->iphdrs.ipv6hdr.nexthdr) {
-                case IPPROTO_TCP:
-                case IPPROTO_UDP:
-                case IPPROTO_ICMPV6:
-                    break;
-                default:
-                    return 1; // unsupported proto
-            }
-
-            // add IPv6 header items to indexer
-            __builtin_memcpy(&indexer.src.in6_u, &nethdrs->iphdrs.ipv6hdr.saddr.in6_u, 4 * sizeof(u32));
-            __builtin_memcpy(&indexer.dst.in6_u, &nethdrs->iphdrs.ipv6hdr.daddr.in6_u, 4 * sizeof(u32));
-            break;
-
-        default:
-            return 1; // verifier
-    }
-
-    net_event_context_t *neteventctx;
-    neteventctx = bpf_map_lookup_elem(cgrpctxmap, &indexer); // obtain event context
-    if (!neteventctx) {
+    u64 cookie = bpf_get_socket_cookie(ctx);
```

Comment:
Since we talked about cleanup of values, did you maybe also investigate using ’BPF_MAP_TYPE_SK_STORAGE’?

I've read a bit about it and it looks promising 

https://ebpf-docs.dylanreimerink.nl/linux/map-type/BPF_MAP_TYPE_SK_STORAGE/

The values are automatically cleaned up. No idea about cloning sockets though, as the values are stored with the socket.

Reply:
Great idea. Yep, there is `BPF_F_CLONE` which does exactly that we need and clones inserted value too. Changed to use BPF_MAP_TYPE_SK_STORAGE